File: Core/Exporters/HtmlExporter.cs
using System.Text;
using Core.Domain;
using Core.Interfaces;

namespace Core.Exporters;

public sealed class HtmlExporter : IQuestionExporter
{
    public void Export(IEnumerable<GeneratedQuestion> questions, Stream output)
    {
        using var writer = new StreamWriter(output, Encoding.UTF8, leaveOpen: true);

        writer.WriteLine("<!DOCTYPE html>");
        writer.WriteLine("<html><head><meta charset=\"utf-8\"><title>Physics Questions</title>");
        writer.WriteLine("<script src=\"https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js\"></script>");
        writer.WriteLine("<style>body{font-family:sans-serif;max-width:800px;margin:0 auto;padding:20px}");
        writer.WriteLine(".question{margin:20px 0;padding:15px;border:1px solid #ddd;border-radius:5px}");
        writer.WriteLine(".answer{color:green;font-weight:bold}.solution{margin-top:10px;color:#555}</style>");
        writer.WriteLine("</head><body><h1>Physics Questions</h1>");

        var index = 1;
        foreach (var q in questions)
        {
            writer.WriteLine("<div class=\"question\">");
            writer.WriteLine($"<p><strong>Question {index}:</strong> {EscapeHtml(q.Text)}</p>");
            writer.WriteLine($"<div class=\"answer\">Answer: {EscapeHtml(q.Answer)}</div>");

            if (q.Choices is { Count: > 0 })
            {
                writer.WriteLine("<div class=\"choices\">");
                foreach (var choice in q.Choices)
                {
                    writer.WriteLine($"<div>- {EscapeHtml(choice)}</div>");
                }
                writer.WriteLine("</div>");
            }

            writer.WriteLine($"<div class=\"solution\">Solution: \\( {q.SolutionLaTeX} \\)</div>");
            writer.WriteLine("</div>");

            index++;
        }

        writer.WriteLine("</body></html>");
    }

    private static string EscapeHtml(string text)
    {
        return text.Replace("&", "&amp;").Replace("<", "&lt;").Replace(">", "&gt;").Replace("\"", "&quot;");
    }
}

File: src/lib/physics/exporters/HtmlExporter.ts
import type {Exporter} from '../contracts';
import type {ExportFormat, GeneratedQuestion} from '../types';
/** Exports questions to a standalone HTML document with KaTeX rendering. */
export class HtmlExporter implements Exporter {
	public readonly format: ExportFormat = 'html';
	public export(questions: GeneratedQuestion[]): string {
		const parts: string[] = [];
		parts.push('<!DOCTYPE html>\n');
		parts.push('<html><head><meta charset="utf-8"><title>Physics Questions</title>\n');
		parts.push('<link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/katex@0.17.0/dist/katex.min.css">\n');
		parts.push('<script src="https://cdn.jsdelivr.net/npm/katex@0.17.0/dist/katex.min.js"></script>\n');
		parts.push('<script src="https://cdn.jsdelivr.net/npm/katex@0.17.0/dist/contrib/auto-render.min.js"></script>\n');
		parts.push('<script>document.addEventListener("DOMContentLoaded",function(){renderMathInElement(document.body,{delimiters:[{left:"$$",right:"$$",display:true},{left:"\\\\[",right:"\\\\]",display:true},{left:"\\\\(",right:"\\\\)",display:false},{left:"$",right:"$",display:false}]});});</script>\n');
		parts.push('<style>body{font-family:sans-serif;max-width:800px;margin:0 auto;padding:20px}');
		parts.push('.question{margin:20px 0;padding:15px;border:1px solid #ddd;border-radius:5px}');
		parts.push('.answer{color:green;font-weight:bold}.solution{margin-top:10px;color:#555}</style>\n');
		parts.push('</head><body><h1>Physics Questions</h1>\n');
		for (const [i, q] of questions.entries()) {
			parts.push('<div class="question">\n');
			parts.push(`<p><strong>Question ${i + 1}:</strong> ${this.escapeHtml(q.text)}</p>\n`);
			parts.push(`<div class="answer">Answer: ${this.escapeHtml(q.answer)}</div>\n`);
			if (q.choices !== undefined) {
				parts.push('<div class="choices">\n');
				for (const choice of q.choices) {
					parts.push(`<div>- ${this.escapeHtml(choice)}</div>\n`);
				}
				parts.push('</div>\n');
			}
			parts.push(`<div class="solution">Solution: ${this.renderSolution(q)}</div>\n`);
			parts.push('</div>\n');
		}
		parts.push('</body></html>');
		return parts.join('');
	}
	private renderSolution(q: GeneratedQuestion): string {
		const latex = q.solution_latex.trim();
		if (latex.length > 0) {
			return `$$${latex}$$`;
		}
		return this.escapeHtml(q.solution_text);
	}
	private escapeHtml(text: string): string {
		return text.replace(/&/g, '&amp;').replace(/</g, '&lt;').replace(/>/g, '&gt;').replace(/"/g, '&quot;').replace(/'/g, '&#39;');
	}
}
